modules/foreman_compute_resource.py

```diff
@@ -98,123 +128,145 @@
+        find_locations,
+        naildown_entity_state,
+        sanitize_entity_dict,
+    )
 
+    from nailgun.entities import (
+        AbstractComputeResource,
+        LibvirtComputeResource,
+        OVirtComputeResource,
+        VMWareComputeResource,
+    )
+
+    has_import_error = False
 except ImportError as e:
     has_import_error = True
     import_error_msg = str(e)
 
-from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.foreman_helper import ForemanEntityAnsibleModule
 
+# This is the only true source for names (and conversions thereof)
+name_map = {
+    'name': 'name',
+    'description': 'description',
+    'provider': 'provider',
+    'organizations': 'organization',
+    'locations': 'location',
+    'datacenter': 'datacenter',
+}
 
-def get_provider_params(provider):
+
+def get_provider_infos(provider):
     provider_name = provider.lower()
 
     if provider_name == 'libvirt':
         return {
-            'credentials': ['url'],
-            'params': ['display_type'],
-            'class': nailgun.entities.LibvirtComputeResource
+            'params': ['url', 'display_type'],
+            'class': LibvirtComputeResource
         }
 
     elif provider_name == 'ovirt':
         return {
-            'credentials': ['url', 'user', 'password'],
-            'params': [],
-            'class': cement.OVirtComputeResource
+            'params': ['url', 'user', 'password', 'datacenter', 'use_v4', 'ovirt_quota'],
+            'class': OVirtComputeResource
         }
 
     elif provider_name == 'vmware':
         return {
-            'credentials': ['url', 'user', 'password'],
-            'params': ['datacenter'],
-            'class': cement.VMWareComputeResource
+            'params': ['url', 'user', 'password', 'datacenter'],
+            'class': VMWareComputeResource
         }
 
     else:
-        return False
+        return {
+            'params': [],
+            'class': AbstractComputeResource
+        }
 
 
-def main(module):
-    if has_import_error:
-        module.fail_json(msg=import_error_msg)
+def main():
+    module = ForemanEntityAnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str', required=True),
+            updated_name=dict(type='str'),
+            description=dict(type='str'),
+            provider=dict(type='str', choices=['vmware', 'libvirt', 'ovirt']),
+            provider_params=dict(type='dict'),
+            locations=dict(type='list'),
+            organizations=dict(type='list'),
+            state=dict(type='str', default='present', choices=['present', 'absent', 'present_with_defaults']),
+            # provider-params (deprecated), use nested keys in provider_params param instead
+            provider_auth=dict(type='dict'),
+            url=dict(type='str'),
+            display_type=dict(type='str'),
+            datacenter=dict(type='str'),
+        ),
+        required_if=(
+            ['state', 'present', ['provider']],
```

Comment:
Would you mind also documenting this? Something like `Required if I(state=present).`